File: python/src/tdcv2/compute/encode.py
"""``<encode as="...">`` — one character turned into a number.

Every check-digit algorithm starts here: a passport number's letters become numbers under base36,
an IBAN's country code becomes two-digit values, a control character is a code point. Each
encoding takes ONE character and returns a STRING, so that the folds downstream consume digits
uniformly whichever encoding produced them.

For base36, ascii and unicode that string is the DECIMAL value of the character, so ``A`` under
base36 is ``"10"``. For hex, binary and octal it is the code point rendered in that base — which is
a different question, and the two are not interchangeable.
"""

from __future__ import annotations

from .value import ComputeError

ENCODINGS = ("base36", "ascii", "unicode", "hex", "binary", "octal")


def code_point_of(ch: str) -> int:
    """The single code point of ``ch``, or an error if it is not exactly one."""
    if len(ch) != 1:
        raise ComputeError(f'<encode>: expected a single character, got "{ch}"')
    return ord(ch)


def encode_char(ch: str, as_what: str) -> str:
    if as_what == "base36":
        return str(_base36_value(ch))
    if as_what == "ascii":
        cp = code_point_of(ch)
        if cp >= 128:
            raise ComputeError(
                f'<encode as="ascii">: "{ch}" is not an ASCII character (code >= 128)'
            )
        return str(cp)
    if as_what == "unicode":
        return str(code_point_of(ch))
    if as_what == "hex":
        return format(code_point_of(ch), "x")
    if as_what == "binary":
        return format(code_point_of(ch), "b")
    if as_what == "octal":
        return format(code_point_of(ch), "o")
    raise ComputeError(
        f'<encode>: unknown encoding "{as_what}" (expected one of {", ".join(ENCODINGS)})'
    )


def _base36_value(ch: str) -> int:
    """0-9 → 0..9, A-Z and a-z → 10..35."""
    cp = code_point_of(ch)
    if 48 <= cp <= 57:
        return cp - 48
    if 65 <= cp <= 90:
        return cp - 65 + 10
    if 97 <= cp <= 122:
        return cp - 97 + 10
    raise ComputeError(f'<encode as="base36">: "{ch}" is not a digit or letter')

File: csharp/Tdcv2/Compute/Encode.cs
using System.Globalization;
using System.Numerics;

namespace Tdcv2.Compute;

/// <summary>
/// <c>&lt;encode as="..."&gt;</c> — one character to a number.
/// </summary>
/// <remarks>
/// <para>
/// The step every alphanumeric check digit needs: an IBAN turns letters into numbers before it takes
/// its mod 97, a vehicle identification number folds letters into digits before weighting.
/// </para>
/// <para>
/// For <c>base36</c>, <c>ascii</c> and <c>unicode</c> the result is the character's <em>decimal</em>
/// value, so <c>A</c> under base36 is the string <c>"10"</c> and a fold then consumes those digits.
/// For <c>hex</c>, <c>binary</c> and <c>octal</c> it is the code point written in that base.
/// </para>
/// <para>
/// A character means one Unicode code point, never one UTF-16 unit — the rest of the layer iterates
/// strings by code point, and an encoding that disagreed would split an emoji in half.
/// </para>
/// </remarks>
public static class Encode
{
    public static string EncodeChar(string ch, string @as) => @as switch
    {
        "base36" => Base36Value(ch).ToString(CultureInfo.InvariantCulture),
        "ascii" => AsciiValue(ch),
        "unicode" => CodePointOf(ch).ToString(CultureInfo.InvariantCulture),
        "hex" => ToBase(CodePointOf(ch), 16),
        "binary" => ToBase(CodePointOf(ch), 2),
        "octal" => ToBase(CodePointOf(ch), 8),
        _ => throw new ComputeError(
            $"<encode as=\"{@as}\">: unknown encoding "
            + "(base36, ascii, unicode, hex, binary, octal)"),
    };

    private static string AsciiValue(string ch)
    {
        int cp = CodePointOf(ch);
        if (cp >= 128)
        {
            throw new ComputeError(
                $"<encode as=\"ascii\">: \"{ch}\" is not an ASCII character (code >= 128)");
        }

        return cp.ToString(CultureInfo.InvariantCulture);
    }

    internal static int CodePointOf(string ch)
    {
        if (ch.Length == 0
            || (char.IsHighSurrogate(ch[0]) ? ch.Length != 2 : ch.Length != 1))
        {
            throw new ComputeError($"<encode>: expected a single character, got \"{ch}\"");
        }

        return char.ConvertToUtf32(ch, 0);
    }

    /// <summary>0-9 to 0..9, and either case of A-Z to 10..35.</summary>
    private static int Base36Value(string ch)
    {
        int cp = CodePointOf(ch);
        if (cp >= '0' && cp <= '9')
        {
            return cp - '0';
        }

        if (cp >= 'A' && cp <= 'Z')
        {
            return cp - 'A' + 10;
        }

        if (cp >= 'a' && cp <= 'z')
        {
            return cp - 'a' + 10;
        }

        throw new ComputeError($"<encode as=\"base36\">: \"{ch}\" is not a digit or letter");
    }

    /// <summary>Lower-case digits, as <c>Integer.toString(n, radix)</c> and <c>n.toString(radix)</c> both write them.</summary>
    private static string ToBase(int value, int radix)
    {
        if (value == 0)
        {
            return "0";
        }

        const string digits = "0123456789abcdefghijklmnopqrstuvwxyz";
        var big = new BigInteger(Math.Abs(value));
        var text = new Stack<char>();
        while (!big.IsZero)
        {
            text.Push(digits[(int)(big % radix)]);
            big /= radix;
        }

        return (value < 0 ? "-" : "") + new string(text.ToArray());
    }
}
